### ISSUE TITLE
Cannot test combobox rendered in portal with react-testing-library (headlessui v2)

### ISSUE THREAD

**Issue by crissadriana** [tailwindlabs/headlessui#3294]
**What package within Headless UI are you using?**
@headlessui/react 

**What version of that package are you using?**

v2.0.4

**Describe your issue**

After upgrading to version 2, the combobox using anchor opens in a portal on the body and the integration tests using react-testing-library are now failing. For example, I'm trying to test a component where I should render the combobox but when I have to check the contents of it, the test is failing because it doesn't find the list of options. I have tried to add the body as a wrapper for the tested component but that doesn't fix it. Do you have any suggestions on that?

**Comment by optimistic-updt** [comment]
+1

**Comment by RobinMalfait** [comment]
Hey!

Can you share a minimal reproduction repo that shows how you are testing the component?

**Comment by optimistic-updt** [comment]
Hey @RobinMalfait, this is one of the most minimal example I can do
made with create-react-app, matching the versions in my package.json

`npm run test` and watch is hang

https://github.com/optimistic-updt/repro-headless-jest-portal

**Comment by JordanVincent** [comment]
Using `userEvent.selectOptions()` worked for me. Note that Headless UI uses [mousedown instead of click](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/combobox/combobox.tsx#L1871) to select options.

```ts
screen.getByRole("input").focus();
const option = (await screen.getByRole("option", { name: "Some option" }));
await userEvent.selectOptions(screen.getByRole("listbox"), selectedOption);
```

**Comment by lukecollins11** [comment]
I’m also facing a similar issue to the above. 
When I am testing with no anchor passed to the options container the test works fine. When I add the anchor back in, the test hangs indefinitely. 

There are warnings in the logs for NaN being an invalid value for the ‘left’ css style property, originating from ‘InternalPortalFn2’

**Comment by lukecollins11** [comment]
> Using `userEvent.selectOptions()` worked for me. Note that Headless UI uses [mousedown instead of click](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/combobox/combobox.tsx#L1871) to select options.
> 
> ```ts
> screen.getByRole("input").focus();
> const option = (await screen.getByRole("option", { name: "Some option" }));
> await userEvent.selectOptions(screen.getByRole("listbox"), selectedOption);
> ```

Are you making use of the anchor property in the options container?

**Comment by JordanVincent** [comment]
Hmm, it works on my end with `anchor`:

```tsx
<ComboboxOptions
    static={true}
    anchor={{ to: "bottom start", gap: 8, padding: 8 }}
>...</ComboboxOptions>
```
Make sure the options are shown. You can inspect the DOM with `screen.debug()`.

**Comment by JordanVincent** [comment]
Ok, so my original solution was flaky. But I was able to fix it that way:

```ts
const requestAnimationFrameMock = jest.spyOn(window, "requestAnimationFrame").mockImplementation(setImmediate as any);
const cancelAnimationFrameMock = jest.spyOn(window, "cancelAnimationFrame").mockImplementation(clearImmediate as any);

screen.getByRole("input").focus();
const option = await screen.getByRole("option", { name: "Some option" });
await userEvent.selectOptions(screen.getByRole("listbox"), selectedOption);
screen.getByRole("input").blur();

requestAnimationFrameMock.mockRestore();
cancelAnimationFrameMock.mockRestore();
```

When `immediate` is set to `true`, [it refocuses the input](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/combobox/combobox.tsx#L1901). This reopens the dropdown but because it's using `requestAnimationTimeframe` it's hard to get the timing right, leading to flakiness. Mocking `requestAnimationTimeframe` fixes the timing issue and calling `.blur()` closes the dropdown. Headless UI's does something similar in [their tests](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/combobox/combobox.test.tsx#L66).

**Comment by RobinMalfait** [comment]
Hey! Small update: this is a bug in Headless UI, and #3357 will solve the issue which means that the reproduction provided here https://github.com/tailwindlabs/headlessui/issues/3294#issuecomment-2187849495 will just work without any changes (apart from a version bump once the PR is ready)

**Comment by RobinMalfait** [comment]
This should be fixed by #3357, and will be available in the next release.

You can already try it using:
- `npm install @headlessui/react@insiders`.

**Comment by Odysseus14** [comment]
> This should be fixed by #3357, and will be available in the next release.
> 
> You can already try it using:
> 
> * `npm install @headlessui/react@insiders`.

This now works for me! Thank you for the quick turnaround!
